app/services/toolbox/google_search.rb

```diff
@@ -0,0 +1,32 @@
+class Toolbox::GoogleSearch < Toolbox
+
+  describe :google_search, <<~S
+    Search Google for the indicated query.
+    Use this to answer questions about current events, look up information, or find answers to questions.
+    Try to use this sparingly; prefer to use the user's memories and the tools you have available to answer questions.
+    When you do use this, try to use exact queries for which you expect to get a definitive answer.
+    When you respond to the user, try to include an answer to the question rather than just a link.
+  S
+  def google_search(query_s:)
+    encoded_query = URI.encode_www_form_component(query_s)
+    response_body = get("https://www.google.com/search?q=#{encoded_query}").get_body
```

Comment:
Instead of get_body, the SDK library is expecting you to pass .params() or .no_params(). Those are the methods that return the body. So try:

```
get(“https://www.google.com/search”).params(q: query_s)
```

I believe that should handle the encoding for you as well and it should return the body. Notably, I wrote this generic SDK because I got tired of reading through API docs for all these different services, and reading through SDK docs wrapping each of the APIs. REST APIs are easy enough these days I just wanted a generic library that let me interact with them. I did not intend it for this specific use-case since google isn’t returning a JSON response, but I think it’ll probably work.

The `params` method ends by calling a helper method called `handle`. You can see it here:
https://github.com/AllYourBot/hostedgpt/blob/b8c9c52dd4916697cdf0332a9cd181946d629834/app/services/sdk/verb.rb#L30

It looks like I wrote it so that if the response is not JSON data then it simply returns the `response` object, so you may actually just want to call it like this:

```
get(“https://www.google.com/search”).params(q: query_s).body
```

Reply:
I originally tried calling it like that, and just tried again but it doesn't work and I think something about the `handle` method or the `OpenData` class has to change.  There's maybe something wrong or something I have to do to the `body` text before I can parse it.
It looks cool but I don't quite understand it yet.  I'll fool around with them a bit.

Reply:
The following works.  Does it look OK?  I'll fool around with it a little more.

```ruby
  def handle(response)
    raise ResponseError.new(response) if !response.status.in? @expected_statuses

    if response.status.between?(200, 299)
      begin
        json = JSON.parse(response.body)
      rescue JSON::ParserError, TypeError => e
        return response
      end
      response.body.presence && OpenData.for(json)
    else
      response
    end
  end
  ```